### PR TITLE
Csr interface organization

### PR DESCRIPTION
**Pull request by cnpetra**
Reorganizes CSR Matrix class by creating a base virtual CSR Matrix interface. Existing Host code is moved to a "Seq" class. Skeleton for the CUDA CSR (that will be using cuSPARSE API and potentially some hand written code) is also created.

Closes issue #432

### CODE REVIEW

**Review by nychiang — APPROVED** [review]
LGTM. Looking for the follow-up PRs.

File: src/LinAlg/hiopMatrixSparseCSRCUDA.hpp [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+  hiopMatrixSparseCSRCUDA(const hiopMatrixSparseCSRCUDA&)
+    : hiopMatrixSparseCSR(0, 0, 0),
+      irowptr_(nullptr),
+      jcolind_(nullptr),
+      values_(nullptr),
+      buf_col_(nullptr),
+      row_starts_(nullptr)
+  {
+    assert(false && "copy constructor disabled");
+  }
```
Comment (ashermancinelli) [inline src/LinAlg/hiopMatrixSparseCSRCUDA.hpp:594]:
```suggestion
  hiopMatrixSparseCSRCUDA(const hiopMatrixSparseCSRCUDA&) = delete;
```

File: src/LinAlg/hiopMatrixSparseCSRSeq.hpp [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+  hiopMatrixSparseCSRSeq(const hiopMatrixSparseCSRSeq&)
+    : hiopMatrixSparseCSR(0, 0, 0),
+      irowptr_(nullptr),
+      jcolind_(nullptr),
+      values_(nullptr),
+      buf_col_(nullptr),
+      row_starts_(nullptr)
+  {
+    assert(false && "copy constructor disabled");
+  }
```
Comment (ashermancinelli) [inline src/LinAlg/hiopMatrixSparseCSRSeq.hpp:592]:
```suggestion
  hiopMatrixSparseCSRSeq(const hiopMatrixSparseCSRSeq&) = delete;
```

**Review by ashermancinelli — APPROVED** [review]

**Review by pelesh — CHANGES_REQUESTED** [review]
Since we are starting fresh, I would suggest using class templates to specify execution policies for CSR matrices. That would eliminate need for separate `CSRseq` class and make the whole code more compact and easier to maintain.

File: src/LinAlg/hiopMatrixSparseCSRSeq.hpp [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+
+  /// @brief Performs a quick check and returns false if the CSR indexes are not ordered
+  bool check_csr_is_ordered();
+  /////////////////////////////////////////////////////////////////////
+  // end of new CSR-specific methods
+  /////////////////////////////////////////////////////////////////////
+
+private:
+  void alloc();
+  void dealloc();
```
Comment (pelesh) [inline src/LinAlg/hiopMatrixSparseCSRSeq.hpp:559]: This is great!

File: src/LinAlg/hiopMatrixSparseCSRSeq.hpp (hunk 2 above)
Comment (pelesh) [reply src/LinAlg/hiopMatrixSparseCSRSeq.hpp:592]: Why is copy constructor disabled? You could make it private and then use it to implement `clone` and `dep_copy` methods, which can be handy.

**Comment by cnpetra** [comment]
> Since we are starting fresh, I would suggest using class templates to specify execution policies for CSR matrices. That would eliminate need for separate `CSRseq` class and make the whole code more compact and easier to maintain.

I am not following... CSRSeq is the reference CPU class. CUDA will be vendor specific. Will have one for AMD. You mean exec policy templates for a portable implementation, likely based on RAJA ?

**Comment by pelesh** [comment]
Consider more compact design using class templates like this:
```c++
template <typename indexT = int, typename scalarT = double, class exec_policy = hiop::Seq>
class hiopMatrixSparseCSR
{
public:
...
 
private:
  indexT rowptrs_;
  indexT colidx_;
  scalarT values_;
}; 

Classes passed as `exec_policy` would contain bare bones (BLAS style) kernels. For each added policy one would implement only those kernels (or substitute third party ones), while using the same class template.

**Comment by pelesh** [comment]
> I am not following... CSRSeq is the reference CPU class. CUDA will be vendor specific. Will have one for AMD. You mean exec policy templates for a portable implementation, likely based on RAJA ?

I need to prototype that in a simple example to communicate it better. It does not need to be based on RAJA.

**Comment by ashermancinelli** [comment]
> > I am not following... CSRSeq is the reference CPU class. CUDA will be vendor specific. Will have one for AMD. You mean exec policy templates for a portable implementation, likely based on RAJA ?
> 
> I need to prototype that in a simple example to communicate it better. It does not need to be based on RAJA.

@cnpetra We may specialize certain methods for the exec policy template parameter, such that the kernels will by default just use sequential or CUDA exec policies in RAJA kernels, but then provide some specialization like so:

```c++
template<typename IndexT, typename ScalarT>
hiop::hiopMatrixSparseCSR<IndexT, ScalarT, hiop::Cuda>::some_specialized_method(...) {
  // some vendor-specific code here
}
```

Apologies if this is not what you were getting at @pelesh. I can provide a small example to show how this might work if that's helpful.

**Comment by ashermancinelli** [comment]
However, if we're comfortable with C++17 (which is supported by every compiler we test with) we can use `if constexpr` which removes the need for as many template shenanigans. [Here's a full example of a matrix class with compile-time specialization](https://godbolt.org/z/s1oWGjKv6).

**Comment by nychiang** [comment]
Thanks @ashermancinelli and @pelesh for the suggestion and example.
I understand how this approach works, but my experience is that it ends up a long file including all the specialized implementations. (For example, all the functions need to have this `if-else statement` and  all the specialized code must be implemented in the `if-else statement`. 
Is there a way to avoid that?

**Comment by pelesh** [comment]
> I understand how this approach works, but my experience is that it ends up a long file including all the specialized implementations. (For example, all the functions need to have this `if-else statement` and all the specialized code must be implemented in the `if-else statement`.
> Is there a way to avoid that?

I believe what I propose in #437 would avoid using conditionals to select specialized source code and if I understand it correctly, @ashermancinelli proposal would do the same. The difference is I am (surprise, surprise) more conservative in using C++ newer features.

File: src/Optimization/hiopDualsUpdater.cpp [hunk 4]
```diff
@@ -413,39 +413,63 @@ bool hiopDualsLsqUpdateLinsysAugSparse::do_lsq_update(hiopIterate& iter,
...
+
+      // This is our first choice on the device.
+      if(linear_solver == "cusolver" || linear_solver == "auto") {
+        hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
         nlp_->log->printf(hovSummary,
                           "LSQ Dual Initialization --- KKT_SPARSE_XDYcYd linsys: using CUSOLVER on device as an "
                           "indefinite solver, size %d (%d cons)\n",
                           n,
                           neq+nineq);
         p->setFakeInertia(neq + nineq);
```
Comment (pelesh) [inline src/Optimization/hiopDualsUpdater.cpp:440]: Remove fake inertia from here. It is not used anyways.
```suggestion
      if(linear_solver == "cusolver" || linear_solver == "auto") {
        hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
        nlp_->log->printf(hovSummary,
                          "LSQ Dual Initialization --- KKT_SPARSE_XDYcYd linsys: using CUSOLVER on device as an "
                          "indefinite solver, size %d (%d cons)\n",
                          n,
                          neq+nineq);
```

File: src/Optimization/hiopKKTLinSysSparse.cpp [hunk 5]
```diff
@@ -572,70 +602,106 @@ namespace hiop
...
+        assert(nullptr == linSys_);
+        if(linear_solver == "cusolver" || linear_solver == "auto") {
+#if defined(HIOP_USE_CUSOLVER)        
+          hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
+          nlp_->log->printf(hovScalars,
+                            "KKT_SPARSE_XDYcYd linsys: alloc CUSOLVER size %d (%d cons) (hybrid)\n",
+                            n,
+                            neq+nineq);
+          p->setFakeInertia(neq + nineq);
+          linSys_ = p;
```
Comment (pelesh) [inline src/Optimization/hiopKKTLinSysSparse.cpp:635]: Remove fake inertia:
```suggestion
          hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
          nlp_->log->printf(hovScalars,
                            "KKT_SPARSE_XDYcYd linsys: alloc CUSOLVER size %d (%d cons) (hybrid)\n",
                            n,
                            neq+nineq);
          linSys_ = p;
```

File: src/Optimization/hiopKKTLinSysSparse.cpp [hunk 6]
```diff
@@ -572,70 +602,106 @@ namespace hiop
...
+        assert(nullptr == linSys_);
+        if(linear_solver == "cusolver" || linear_solver == "auto") {
+#if defined(HIOP_USE_CUSOLVER)        
+          hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
+          nlp_->log->printf(hovScalars,
+                            "KKT_SPARSE_XDYcYd linsys: alloc CUSOLVER size %d (%d cons) (gpu)\n",
+                            n,
+                            neq+nineq);
+          p->setFakeInertia(neq + nineq);
+          linSys_ = p;
```
Comment (pelesh) [inline src/Optimization/hiopKKTLinSysSparse.cpp:698]: Remove fake inertia:
```suggestion
          hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
          nlp_->log->printf(hovScalars,
                            "KKT_SPARSE_XDYcYd linsys: alloc CUSOLVER size %d (%d cons) (gpu)\n",
                            n,
                            neq+nineq);
          linSys_ = p;
```

**Comment by pelesh** [comment]
When running example 7:
```shell
jsrun -n 1 -g 1 ./src/Drivers/nlpSparse_ex7.exe 500 -cusolver -inertiafree -selfcheck
```
for condensed model I get following output:
```shell
LSQ Dual Updater --- KKT_SPARSE_XDYcYd linsys: using MA57 on CPU(!!!) size 1502 (501 cons)
Input Error: Incorrect objective type.
```
After that computation seems to run fine using Cholesky solver module:
```shell
...
[Warning] hiopLinSolverCholCuSparse: the 499th pivot is <=1.00000e-24
[Warning] solve compressed high absolute resid norm (= 8.45063e-02)
   7  2.1799950e+02 4.050e-13  2.726e-01  -1.40  9.321e-01  1.000e+00  1(s)
...
```
It is unclear to me what solver is invoked when, but it looks like first solve is done with MA57 and subsequent with Cholesky. Is that the idea here?

**Comment by nychiang** [comment]
@pelesh In the first iteration, the dual variables are computed by a least-square subproblem. It is passed to MA57, and failed Not sure why the error msg is "Incorrect objective type", but it happens if the matrix is singular (Ex7 has rank deficient constraints). When the least-square initialization fails, HiOp set all the duals to zero and continue the iterates.
Therefore, you may want to set `duals_init = zero` to ignore the lsq initialization, and just use cuSolvers in HiOp.

**Comment by pelesh** [comment]
> When the least-square initialization fails, HiOp sets all the duals to zero and continue the iterates.
> Therefore, you may want to set `duals_init = zero` to ignore the lsq initialization, and just use cuSolvers in HiOp.

@nychiang, you are right on the money. Setting `duals_init` to `zero` seems to fix the logic of selecting Cholesky solver, as well. It seems that using `lsq` for duals initialization breaks the solver selection logic when cuSolver (Cholesky or LU) is enabled. 

CC @kswirido

**Comment by cnpetra** [comment]
shoot, @pelesh , I deleted you reply, please repost

**Comment by cnpetra** [comment]
> @nychiang, you are right on the money. Setting duals_init to zero seems to fix the logic of selecting Cholesky solver, as well. It seems that using lsq for duals initialization breaks the solver selection logic when cuSolver (Cholesky or LU) is enabled.
> 
> CC @kswirido[kswirydo](https://github.com/kswirydo)

Just to clarify, the Cholesky solver should be selected in the second test in ex7 driver, independent of the other options (assuming CUDA is present and HIOP_SPARSE is on)...

**Comment by pelesh** [comment]
> Just to clarify, the Cholesky solver should be selected in the second test in ex7 driver, when cusolver is required. the LU solver should solve the first test in the driver

Agreed!

**Comment by cnpetra** [comment]
> > Just to clarify, the Cholesky solver should be selected in the second test in ex7 driver, when cusolver is required. the LU solver should solve the first test in the driver
> 
> Agreed!

does it work like that for both when duals initialization set to lsq and to zero? or you have seen otherwise?

**Comment by pelesh** [comment]
> does it work like that for both when duals initialization set to lsq and to zero? or you have seen otherwise?

In #427 I find that @nychiang's suggestion to initialize duals with zero fixes issues with both cuSolver LU (first test) and Cholesky (secod test). In particular, in second test, it seems that setting duals initialization to zero also fixes the issue where LU factorization is called instead of Cholesky. It looks to me that requesting `lsq` initialization of duals messes up the solver selection logic.

**Comment by cnpetra** [comment]
> > does it work like that for both when duals initialization set to lsq and to zero? or you have seen otherwise?
> 
> In #427 I find that @nychiang's suggestion to initialize duals with zero fixes issues with both cuSolver LU (first test) and Cholesky (secod test). In particular, in second test, it seems that setting duals initialization to zero also fixes the issue where LU factorization is called instead of Cholesky. It looks to me that requesting `lsq` initialization of duals messes up the solver selection logic.

I just checked and this PR fixes that.

File: src/Optimization/hiopDualsUpdater.cpp (hunk 4 above)
Comment (cnpetra) [reply src/Optimization/hiopDualsUpdater.cpp:440]: it turns out that cuSOLVER LU still relies on the fake inertia in this branch and stops working if I remove it.

File: src/Drivers/nlpSparse_ex6_driver.cpp [hunk 7]
```diff
@@ -176,6 +176,7 @@ int main(int argc, char **argv)
     nlp.options->SetStringValue("linear_solver_sparse", "pardiso");
   }
   if(use_cusolver) {
+    nlp.options->SetStringValue("linsol_mode", "speculative");
     nlp.options->SetStringValue("linear_solver_sparse", "cusolver");
     nlp.options->SetStringValue("compute_mode", "hybrid");
```
Comment (pelesh) [inline src/Drivers/nlpSparse_ex6_driver.cpp:181]: Need to enable `inertia_free` for cuSolver:
```suggestion
    nlp.options->SetStringValue("linsol_mode", "speculative");
    nlp.options->SetStringValue("linear_solver_sparse", "cusolver");
    nlp.options->SetStringValue("compute_mode", "hybrid");
    nlp.options->SetStringValue("fact_acceptor", "inertia_free");
```

Comment (pelesh) [reply src/Drivers/nlpSparse_ex6_driver.cpp:181]: Or merge with `cusolver-inertia-free-dev` :)

Comment (cnpetra) [reply src/Drivers/nlpSparse_ex6_driver.cpp:181]: > Or merge with `cusolver-inertia-free-dev` :)

I hacked it for now
https://github.com/LLNL/hiop/blob/22ad66217e79c3d4086cc2c79e0cf908e722c681/src/Optimization/hiopKKTLinSysSparse.cpp#L638

Comment (pelesh) [reply src/Drivers/nlpSparse_ex6_driver.cpp:181]: No, the issue is with sparse ex6, not "fake inertia".

Comment (cnpetra) [reply src/Drivers/nlpSparse_ex6_driver.cpp:181]: it does not work with 
`
nlp.options->SetStringValue("fact_acceptor", "inertia_free");
`

File: src/Optimization/hiopDualsUpdater.cpp [hunk 8]
```diff
@@ -413,39 +413,63 @@ bool hiopDualsLsqUpdateLinsysAugSparse::do_lsq_update(hiopIterate& iter,
...
+      //under compute mode gpu, at this point we don't have a sparse linear solver 
+      if(compute_mode == "gpu") {
+        if(linear_solver == "auto") {
+          assert("HiOp was not built with a sparse GPU and cannot fullfil the requirement of the option "
+                 "'duals_init_linear_solver_sparse'. Either build with a supported GPU sparse solver or "
+                 "change compute mode to hybrid, which will allow using a CPU sparse solver.");
+        } else {
+          assert("Impossible to deploy the (CPU?) sparse linear solver specified by option "
+                 "'duals_init_linear_solver_sparse' under gpu compute mode. Either build with a supported "
+                 "GPU sparse solver or change compute mode to hybrid, which will allow using CPU sparse solvers.");
```
Comment (nychiang) [inline src/Optimization/hiopDualsUpdater.cpp:453]: Two comments:
1. when we use this function, `do_lsq_update`, to initialize dual variables, we can return `false` whenever assertion fails. Then the outer loop ([here](https://github.com/LLNL/hiop/blob/e29104a18e380f9a537ddc2a960a4e34a95398cf/src/Optimization/hiopDualsUpdater.hpp#L139-L159)) will set duals to `zero`
2. this function can also be used in standard ipm iterations, e.g., solve the duals for quasi-newton method, and hence the error message may need to be modified. That is, the problem is not about `duals_init_linear_solver_sparse`, but `linear_solver_sparse` when `duals_update_type = lsq`.

Comment (cnpetra) [reply src/Optimization/hiopDualsUpdater.cpp:453]: good point, will revisit

File: src/Drivers/nlpSparse_ex6_driver.cpp (hunk 7 above)
Comment (pelesh) [reply src/Drivers/nlpSparse_ex6_driver.cpp:181]: I can run it on my machines without calls to fake inertia, see #440.

**Review by pelesh — APPROVED** [review]
The code builds correctly and all tests pass. The approach to add CSR matrices is consistent with the rest of HiOp architecture.

There is a large number of warnings when built in "developer's mode", which need to be addressed eventually, but that's beyond the scope of this PR.

**Pull request merged**